sequencer/src/block2.rs

```diff
@@ -0,0 +1,144 @@
+// Question: jellyfish uses `ark_std` everywhere instead of `std`.
+// Is there a chance of badness if we (as a downstream user) use `std`?
+use std::mem::size_of;
+
+#[allow(dead_code)] // TODO temporary
+pub struct BlockPayload {
+    payload: Vec<u8>,
+    // tx_ranges: Vec<Range<u32>>, // not convinced we need this
```

Comment:
Yeah, I agree. We originally had the idea for this when we were planning on putting each transactions size with the transaction, so finding the transaction boundaries required a linear scan. Now with the transactions table this is no longer necessary.

Reply:
We should store the VID commitment in here, though, so that we can retrieve it in constant time.

Reply:
> We should store the VID commitment in here, though, so that we can retrieve it in constant time.

How would that work with `build()`? Seems like `build()` would need to compute the VID commitment, but I've been trying hard to avoid unnecessary computations of the VID commitment. (We could make the commitment an `Option` but I hope I don't need to argue against that.)

Note that the VID commitment is not needed to make namespace/tx proofs; it's needed only to verify such a proof.

Perhaps the VID commitment could instead be given to us as part of the metadata / header.

Reply:
Currently in HotShot the `BlockPayload` trait has a `commit()` function, and it's understood that calling `build()` computes the commitment, and thereafter you keep the thing you built and it's cheap to call `commit()`. But maybe it makes more sense to have HotShot separately keep track of the commitment, so the `BlockPayload` implementation doesn't have to worry about it. After all, HotShot knows best when it needs to compute a commitment and when it needs to remember that commitment. @shenkeyao

Reply:
> But maybe it makes more sense to have HotShot separately keep track of the commitment, so the `BlockPayload` implementation doesn't have to worry about it. After all, HotShot knows best when it needs to compute a commitment and when it needs to remember that commitment.

Strongly agree. We're already taking great pains in hotshot to minimize where we compute the VID commitment. I'd hate to waste that effort by computing it again inside `build()`.

Reply:
This makes sense! Let me check the codebase to see if removing `commit()` is doable. Just to clarify, `BlockPayload` actually doesn't have a `commit()` function, but it has the `Committable` bound which has `commit()`, so we need to remove the bound.

Reply:
I think we still need a function to compute the commitment. If we are trying to avoid the computation in `build()`, I think we could add a `vid_commitment` function to `Payload` but unlike `commit()` which takes `&self` as input, it takes the encoded transactions. In this way, the payload build and the commitment computation are separated.

Reply:
Why do we need a trait function for this? The thing that's app-specific is `encode()`, but once you have a flat list of bytes you just need to call `disperse` or `commit_only`, and it should actually be HotSHot's responsibility to (a) know which one to call when and (b) set up the VID scheme with the proper parameters

Reply:
Agreed. I just created an [issue](https://github.com/EspressoSystems/HotShot/issues/2072) for this. Would like to do it in a separate PR.

Reply:
Yeah that makes sense